ssl/quic/quic_impl.c

```diff
@@ -9,271 +9,1244 @@
+    size_t actual_written = 0;
+
+    if (!ossl_quic_channel_is_active(args->qc->ch))
+        /* If connection is torn down due to an error while blocking, stop. */
+        return -2;
+
+    if (!ossl_quic_sstream_append(args->qc->stream0->sstream,
+                                  args->buf, args->len, &actual_written))
+        return -2;
+
+    quic_post_write(args->qc, actual_written > 0, 0);
+
+    args->buf           += actual_written;
+    args->len           -= actual_written;
+    args->total_written += actual_written;
+
+    if (actual_written == 0)
+        /* Written everything, done. */
+        return 1;
+
+    /* Not written everything yet, keep trying. */
+    return 0;
+}
+
+static int quic_write_blocking(QUIC_CONNECTION *qc, const void *buf, size_t len,
+                               size_t *written)
 {
+    int res;
+    struct quic_write_again_args args;
+    size_t actual_written = 0;
+
+    /* First make a best effort to append as much of the data as possible. */
+    if (!ossl_quic_sstream_append(qc->stream0->sstream, buf, len,
+                                  &actual_written)) {
+        /* Stream already finished or allocation error. */
+        *written = 0;
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+    }
+
+    quic_post_write(qc, actual_written > 0, 1);
+
+    if (actual_written == len) {
+        /* Managed to append everything on the first try. */
+        *written = actual_written;
+        return 1;
+    }
+
+    /*
+     * We did not manage to append all of the data immediately, so the stream
+     * buffer has probably filled up. This means we need to block until some of
+     * it is freed up.
+     */
+    args.qc             = qc;
+    args.buf            = (const unsigned char *)buf + actual_written;
+    args.len            = len - actual_written;
+    args.total_written  = 0;
+
+    res = block_until_pred(qc, quic_write_again, &args, 0);
+    if (res <= 0) {
+        if (!ossl_quic_channel_is_active(qc->ch))
+            return QUIC_RAISE_NON_NORMAL_ERROR(qc, SSL_R_PROTOCOL_IS_SHUTDOWN, NULL);
+        else
+            return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+    }
+
+    *written = args.total_written;
     return 1;
 }
 
-const SSL_CIPHER *ossl_quic_get_cipher(unsigned int u)
+static void aon_write_begin(QUIC_CONNECTION *qc, const unsigned char *buf,
+                            size_t buf_len, size_t already_sent)
+{
+    assert(!qc->aon_write_in_progress);
+
+    qc->aon_write_in_progress = 1;
+    qc->aon_buf_base          = buf;
+    qc->aon_buf_pos           = already_sent;
+    qc->aon_buf_len           = buf_len;
+}
+
+static void aon_write_finish(QUIC_CONNECTION *qc)
+{
+    qc->aon_write_in_progress   = 0;
+    qc->aon_buf_base            = NULL;
+    qc->aon_buf_pos             = 0;
+    qc->aon_buf_len             = 0;
+}
+
+static int quic_write_nonblocking_aon(QUIC_CONNECTION *qc, const void *buf,
```

Comment:
aon?

Reply:
All-or-nothing semantics, the opposite of partial write mode. Added a comment.

Reply:
Yeah....I eventually figured out what it stood for. A comment for others would be useful (and we really ought to have a glossary....numerous times I've run into an acronym I used to know, but now can't remember what it stood for!).